File: task4/task4_java/Potion.java
/*
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Lun Yin Fung
 * Student ID : 1155092566
 * Email Addr : 1155092566@link.cuhk.edu.hk
 */

 public class Potion{

        private static final int HEAL_CAP = 20;
	private Pos pos;
        private int index;
        private Map map;
        private String name;
        private int power;// the healing power
        
        public Potion(int posx, int posy, int index, Map map) {
                this.pos = new Pos(posx, posy);
                this.index = index;
                this.map = map;
                this.name = "P" + Integer.toString(index);
                this.power = TheJourney.rand.nextInt(HEAL_CAP - 5) + 5;
        }

        public boolean actionOnWarrior(Warrior warrior) {
                warrior.talk("Very good, I got additional healing potion " + this.getName() + ".");
                warrior.increaseHealth(this.getPower());
                if (warrior.getHealth() > 40){
                        // prevent the warrior health exceed the health cap
                        warrior.setHealth(40);
                }
                this.map.deleteTeleportableObj(this);
                // return True means the warrior will occupy this position
                return true;
        }

        public void teleport() {
                // the teleport function of the Potion
                this.map.setLand(this.getPos(), null);
                Pos newPos = this.map.getUnOccupiedPosition();
                this.pos.setPos(newPos.getX(), newPos.getY());
                this.map.setLand(this.getPos(), this);
        }

        public Pos getPos() {
		return pos;
        }
        
	public void setPos(Pos pos) {
		this.pos = pos;
        }
        
	public String getName() {
		return name;
        }
        
        public int getPower() {
		return power;
        }
        
}

File: task4/potion.py
import random
from pos import Pos

class Potion:

    _HEAL_CAP = 20

    def __init__(self, posx, posy, index, map):
        self._map = map
        self._pos = Pos(posx,posy)
        self._index = index
        self._name = "P" + str(index)
        self._power = random.randint(0,self._HEAL_CAP - 6) + 5

    @property
    def power(self):
        return self._power

    @property
    def pos(self):
        return self._pos

    @pos.setter
    def pos(self, pos):
        self._pos = pos

    @property
    def name(self):
        return self._name

    def action_on_warrior(self, warrior):
        warrior.talk("Very good, I got additional healing potion %s." % self.name)
        warrior.increase_health(self.power)
        if warrior.health > 40:
            warrior.health = 40
        self._map.delete_teleportable_obj(self)
        # return True means the warrior will occupy this position
        return True

    def teleport(self):
        # print("Teleporting "+self.name)
        self._map.lands = (self.pos, None)
        self.pos = self._map.get_un_occupied_position()
        self._map.lands = (self.pos, self)
        # print("Teleport finish "+self.name)
